strawberryfields/decompositions.py

```diff
@@ -79,36 +107,40 @@ def takagi(N, tol=1e-13, rounding=13):
     return rl, U
 
 
-def graph_embed(mat, max_mean_photon=1.0, make_traceless=True, tol=1e-6):
-    r""" Given an symmetric adjacency matrix (in general, with arbitrary complex off-diagonal and
-    real diagonal entries),
-    it returns the squeezing parameters and interferometer necessary for
+def graph_embed(A, max_mean_photon=1.0, make_traceless=True, tol=1e-6):
+    r"""Embed a graph into a Gaussian state.
+
+    Given a graph in terms of a symmetric adjacency matrix
+    (in general with arbitrary complex off-diagonal and real diagonal entries),
+    returns the squeezing parameters and interferometer necessary for
     creating the Gaussian state whose off-diagonal parts are proportional to that matrix.
 
+    Uses :func:`takagi`.
+
     Args:
-        mat (array): square symmetric complex (or real or integer) array representing a (weighted) adjacency matrix of a graph
-        max_mean_photon (float): threshold value. It guarantees that the mode with
+        A (array[complex]): square, symmetric (weighted) adjacency matrix of the graph
+        max_mean_photon (float): Threshold value. It guarantees that the mode with
             the largest squeezing has ``max_mean_photon`` as the mean photon number
-            i.e., :math:`sinh(r_{max})^2 == max_mean_photon`
-        make_traceless (boolean): removes the trace of the input matrix, by performing the transformation
+            i.e., :math:`sinh(r_{max})^2 ==` max_mean_photon.
+        make_traceless (bool): Removes the trace of the input matrix, by performing the transformation
             :math:`\tilde{A} = A-\mathrm{tr}(A) \I/n`. This may reduce the amount of squeezing needed to encode
             the graph.
-        tol (float): the tolerance used when checking if the input matrix is symmetric: :math:`|mat-mat^T| < tol`
+        tol (float): Tolerance used when checking if the input matrix is symmetric: :math:`|A-A^T| <` tol.
 
     Returns:
-        tuple(array, array): tuple containing the squeezing parameters of the input
-        state to the interferometer, and the unitary matrix representing the interferometer
+        tuple[array, array]: squeezing parameters of the input
+            state to the interferometer, and the unitary matrix representing the interferometer
     """
-    (m, n) = mat.shape
+    (m, n) = A.shape
 
     if m != n:
-        raise ValueError("The Matrix is not square")
+        raise ValueError("The matrix is not square.")
 
-    if np.linalg.norm(mat-np.transpose(mat)) >= tol:
-        raise ValueError("The input matrix is not symmetric")
+    if np.linalg.norm(A-np.transpose(A)) >= tol:
+        raise ValueError("The matrix is not symmetric.")
 
     if make_traceless:
-        A = mat - np.trace(mat)*np.identity(n)/n
+        A = A - np.trace(A)*np.identity(n)/n
```

Comment:
I understand now. I didn't know that in python numpy arrays are not passed by reference. The way it was written before was assuming that they were passed by reference and thus I was trying to prevent the matrix the user passes from being overwritten. I understand this is sub optimal, but fail to see why it is a bug.

Reply:
Yeah, Python can be occasionally surprising if you come from a C/C++ background. What happens here (as far as I know :) is that the subtraction creates a new array object that then gets bound to the local symbol A by the assignment, which releases the previous binding. 

The bug: previously, if make_traceless was False, nothing would be assigned to A, and the takagi call below would throw an error.